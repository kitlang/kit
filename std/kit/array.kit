/**
 * Array is a heap-allocated, length-prefixed, fixed-length collection type.
 */
abstract Array[T]: Slice[T] {
    public static function new(allocator: Box[Allocator], length: Size): Self {
        var data: Ptr[T] = allocator.calloc(sizeof T * length);
        return struct Self {
            length,
            data,
        };
    }

    public function blit(other: Ptr[Array[T]], start: Size, length: Size): Void {
        memcpy(&(other.data[start]) as Ptr[Void], this.data as Ptr[Void], length * sizeof T);
    }

    public function copy(allocator: Box[Allocator]): Array[T] using implicit allocator {
        var a = Self.new(this.length);
        this.blit(a, 0, this.length);
        return a;
    }

    public function free(allocator: Box[Allocator]) {
        allocator.free(this.data);
    }
}
