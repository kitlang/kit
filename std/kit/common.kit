#[builtin] abstract Bool: Void;

#[builtin] abstract CArray[T, $N = 0]: Ptr[T] {
    rules {
        ($this.length) => N;
        (for $ident in $this {$e}) => {
            for __i in 0 ... N {
                var $ident: T = $this[__i];
                {$e}
            }
        }
    }
}
#[builtin] abstract Box[T];

#[builtin] #[promote] abstract Const[T]: T;

#[promote] abstract ReadOnly[T]: T {
    rules {
        ($this.$field = $e) => {1 / 0;}
    }
}

#[promote] #[demote] abstract CString: Ptr[Char] {
    rules {
        ($this.length) => strlen($this);
        ($this == ${other: CString}) => strcmp($this, $other) == 0;
        ($this != ${other: CString}) => strcmp($this, $other) != 0;
        (for $ident in $this {$e}) => {
            var __i: Ptr[Char] = $this;
            while __i != null && (*__i) != 0 {
                var $ident: Char = *(__i++);
                {$e}
            }
        }
    }

    public static function alloc(allocator: Box[Allocator], chars: Size): CString {
        return allocator.alloc(chars + 1);
    }

    public function copy(allocator: Box[Allocator]): CString {
        var new = allocator.alloc(this.length + 1);
        strcpy(new, this);
        return new;
    }

    public function startsWith(other: CString): Bool {
        var len1 = this.length;
        var len2 = other.length;
        return len1 >= len2 && strncmp(this, other, len2) == 0;
    }

    public function endsWith(other: CString): Bool {
        var len1 = this.length;
        var len2 = other.length;
        return len1 >= len2 && strcmp(&this[len1 - len2], other) == 0;
    }
}

trait Enum {
    public static function name(): CString;
    public static function members(): Slice[CString];
}

trait Struct {
    public static function name(): CString;
    public static function fields(): Slice[CString];
}

trait Union {
    public static function name(): CString;
    public static function fields(): Slice[CString];
}
