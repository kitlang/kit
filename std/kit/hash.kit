 /**
 * Hashable trait must be implemented for any key in the hashmap,
 * as the hash function is used to determine it's place in the map.
 */
trait Hashable {
    function hash(): Int;
}

enum HashCellState {
    Empty;
    Deleted;
    Occupied;
}

implement Hashable for Bool {
    function hash(): Int {
        return if this then 1 else 0;
    }
}

implement Hashable for CString {
    function hash(): Int {
        var hash = 5381;
        for char in this {
            hash = (hash << 5) + hash + char;
        }
        return hash;
    }
}

implement Hashable for Int {
    function hash(): Int {
        return this;
    }
}

implement Hashable for Char {
    function hash(): Int {
        return this;
    }
}

implement Hashable for String {
    function hash(): Int {
        var hash = 5381;
        for char in this {
            hash = (hash << 5) + hash + char;
        }
        return hash;
    }
}