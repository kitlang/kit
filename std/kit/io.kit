/**
 * An abstract reader.
 */
trait Reader {
    public function readBytes(buf: Ptr[Void], bytes: Size): Size;
    public function close(): Void;
}

/**
 * An abstract writer.
 */
trait Writer {
    public function writeBytes(buf: Ptr[Void], bytes: Size): Size;

    public function write(v: Box[Writable]): Size {
        return this.writeBytes(v.ptr(), v.size());
    }

    public function flush(): Void {}
    public function close(): Void {}
}

trait Writable {
    public function ptr(): Ptr[Void];
    public function size(): Size;
}

implement Writable for CString {
    public function ptr(): Ptr[Void] {
        return this;
    }

    public function size(): Size {
        return this.length + 1;
    }
}
