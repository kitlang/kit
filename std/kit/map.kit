/**
 * Hashable trait must be implemented for any key in the hashmap,
 * as the hash function is used to determine it's place in the map.
 */
trait Hashable {
    function hash(): Int;
}

/**
 * Generic HashMap implementation, requiring the key to implement the Hashable trait.
 * Relies on == operator for comparisons, so you may need to overwrite it in custom types.
 */
struct Map[K: Hashable, V] {
    static const internalArrayThreshold: Float = 0.7;

    var length: Size = 0;
    var allocator: Box[Allocator];
    var internalArray: Array[KeyValuePair[K, V]];

    public static function new(allocator: Box[Allocator], capacity: Int): Map[K, V] using implicit allocator {
        var internalArray: Array[KeyValuePair[K, V]] = Array.new(capacity);
        return struct Self {
            allocator,
            internalArray,
        };
    }

    public function put(key: K, value: V): Ptr[V] {
        var ref = this.putRef(key);
        *ref = value;
        return ref;
    }

    public function putRef(key: K): Ptr[V] {
        var alreadyExists = this.exists(key);
        if ((this.length as Float) / this.internalArray.length) >= Self.internalArrayThreshold {
            this.resize();
        }
        var index = this.findLocation(key);
        this.internalArray[index] = struct KeyValuePair[K, V] {
            key: key,
            // value: value,
            isActive: true,
        };
        if !alreadyExists {
            ++this.length;
        }
        return &this.internalArray[index].value;
    }

    public function get(key: K): Option[V] {
        var index = this.findLocation(key);
        if this.internalArray[index].isActive {
            return Some(this.internalArray[index].value);
        }
        return None;
    }

    public function ref(key: K): Option[Ptr[V]] {
        var index = this.findLocation(key);
        if this.internalArray[index].isActive {
            return Some(&this.internalArray[index].value);
        }
        return None;
    }

    public function remove(key: K): Void {
        if this.exists(key) {
            this.internalArray[this.findLocation(key)].isActive = false;
            --this.length;
        }
    }

    public function exists(key: K): Bool {
        var index = this.findLocation(key);
        if this.internalArray[index].isActive && this.internalArray[index].key == key {
            return true;
        }
        return false;
    }

    public function keys(): Array[K] {
        var returnArray: Array[K] = Array.new(this.length);
        var j = 0;
        for slot in this.internalArray {
            if slot.isActive {
                returnArray[j++] = slot.key;
            }
        }
        return returnArray;
    }

    public function free(): Void {
        this.internalArray.free();
    }

    public function clear(): Void {
        this.internalArray.clear();
        this.length = 0;
    }

    function findLocation(key: K): Int {
        var hash = key.Hashable.hash() % this.internalArray.length;
        while this.internalArray[hash].isActive && this.internalArray[hash].key != key {
            hash = (hash + 1) % this.internalArray.length;
        }
        return hash;
    }

    function resize(): Void using implicit this.allocator {
        var old = this.internalArray;
        this.internalArray = Array.new(this.internalArray.length * 2);
        for i in 0 ... old.length {
            if old[i].isActive {
                this.put(old[i].key, old[i].value);
                --this.length;
            }
        }
        old.free();
    }

    rules {
        ($this[$k] = $v) => $this.put($k, $v);
        ($this[$k]) => $this.get($k).unwrap();

        (for $ident in $this {$e}) => {
            var __length = $this.internalArray.length;
            for __i in 0 ... __length {
                var __slot = $this.internalArray[__i];
                if __slot.isActive {
                    var $ident = __slot.key;
                    {$e}
                }
            }
        }
    }
}

struct KeyValuePair[K: Hashable, V] {
    public var key: K;
    public var value: V;
    public var isActive: Bool = false;
}

// abstract WeakMap[K, V]: Map[K, Shared[V]] {
//     public static function new(allocator: Box[Allocator], capacity: Int): Self using implicit allocator {
//         return Map[K, Shared[V]].new(capacity) as WeakMap[K, V];
//     }

//     public function get(key: K): Option[Shared[V]] {
//         match this.get(key) {
//             Some(ptr) => {
//                 if ptr.active() {
//                     return Some(ptr);
//                 } else {
//                     this.remove(key);
//                     return None;
//                 }
//             }
//         }
//     }

//     public function retain(key: K): Option[Shared[V]] {
//         match this.get(key) {
//             Some(ptr) => {
//                 return Some(ptr.ref());
//             }
//             default => {
//                 return None;
//             }
//         }
//     }

//     public function release(key: K): Bool {
//         match this.get(key) {
//             Some(ptr) => {
//                 return ptr.release();
//             }
//             default => {
//                 return false;
//             }
//         }
//     }
// }

implement Hashable for Bool {
    function hash(): Int {
        return if this then 1 else 0;
    }
}

implement Hashable for CString {
    function hash(): Int {
        var hash = 5381;
        for char in this {
            hash = (hash << 5) + hash + char;
        }
        return hash;
    }
}

implement Hashable for Int {
    function hash(): Int {
        return this;
    }
}
