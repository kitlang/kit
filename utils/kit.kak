# http://www.kitlang.org
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾

# Detection
# ‾‾‾‾‾‾‾‾‾

hook global BufCreate .*\.(kit) %{
    set-option buffer filetype kit
}

# Highlighters
# ‾‾‾‾‾‾‾‾‾‾‾‾

add-highlighter shared/kit regions
add-highlighter shared/kit/code default-region group
add-highlighter shared/kit/string  region '"' (?<!\\)(\\\\)*"         fill string
add-highlighter shared/kit/comment region '//' '$'                     fill comment

add-highlighter shared/kit/code/ regex %{\b(true|false|None|Some)\b} 0:value
add-highlighter shared/kit/code/ regex \b(struct|enum|union|abstract|function|implement|specialize|trait|typedef|var|private|public|static|const|inline|using|return|throw|this|for|as|in|if|then|else|do|while|match|default|unsafe|rule|rules|implicit|sizeof|import|print|printf)\b 0:keyword
add-highlighter shared/kit/code/ regex \b(CArray|CString|Char|Ptr|Size|Short|Long|Byte|Float|Double|UInt8|UInt16|UInt32|UInt64|Int|Int8|Int16|Int32|Int64|Float32|Float64|)\b 0:type

# Initialization
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾

hook -group kit-highlight global WinSetOption filetype=kit %{ add-highlighter window/kit ref kit }
hook -group kit-highlight global WinSetOption filetype=(?!kit).* %{ remove-highlighter window/kit }
