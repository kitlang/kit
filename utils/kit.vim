" Vim syntax file
"
" Language: Kit
" Last Change: 2018-11-08

if exists("b:current_syntax")
  finish
endif
let b:current_syntax = "kit"


syn region kitCommentOneLine start="//" end="$"
syn region kitCommentMultiLine start="/\*" end="\*/"

syn keyword kitConstant null empty this Self Some None Cons Empty
syn region kitStringOneLine start=/"/ skip=/\\"/ end=/"/ oneline
syn region kitStringMultiLine start=/"""/ end=/"""/
syn match kitCharacter /c'.'/
" syn match kitNumber ...
syn keyword kitBoolean true false
" syn match kitFloat ...

" syn match kitIdentifier ...
" syn match kitFunction ...

" syn keyword kitStatement ...
syn keyword kitConditional if then else match
syn keyword kitRepeat for do while
syn keyword kitLabel default
syn keyword kitOperator sizeof
syn keyword kitKeyword function trait implement return as break defer
syn keyword kitException throw

" syn keyword kitPreProc ...
syn keyword kitInclude include import
" syn keyword kitDefine ...
syn keyword kitMacro rules using rule implicit
" syn keyword kitPreCondit ...

syn keyword kitType Int8 Int16 Short Int32 Int64 Long
syn keyword kitType Uint8 Byte Uint16 Unt32 Uint64
syn keyword kitType Float32 Float Float64 Double
syn keyword kitType Void Bool Char Int Size Ptr
syn keyword kitType Numeric Integral NumericMixed
syn keyword kitType List CArray CString Box
syn keyword kitType Iterable Iterator Option
syn keyword kitType Allocator LinearAllocator StackAllocator
syn keyword kitType Path
syn keyword kitStorageClass public static
syn keyword kitStorageClass var const
" syn keyword kitStorageClass register volatile
syn keyword kitStructure struct union enum abstract
syn keyword kitTypedef typedef specialize

" syn keyword kitSpecial
" syn keyword kitSpecialChar
" syn keyword kitTag
" syn keyword kitDelimiter
" syn keyword kitSpecialComment
" syn keyword kitDebug

" syn keyword kitUnderlined
" syn keyword kitIgnore
" syn keyword kitError
" syn keyword kitTodo


hi def link kitCommentOneLine Comment
hi def link kitCommentMultiLine Comment

hi def link kitConstant Constant
hi def link kitStringOneLine String
hi def link kitStringMultiLine String
hi def link kitCharacter Character
" hi def link kitNumber Number
hi def link kitBoolean Boolean
" hi def link kitFloat Float

" hi def link kitIdentifier Identifier
" hi def link kitFunction Function

" hi def link kitStatement Statement
hi def link kitConditional Conditional
hi def link kitRepeat Repeat
hi def link kitLabel Label
hi def link kitOperator Operator
hi def link kitKeyword Keyword
hi def link kitException Exception

" hi def link kitPreProc Preproc
hi def link kitInclude Include
" hi def link kitDefine Define
hi def link kitMacro Macro
" hi def link kitPreCondit PreCondit

hi def link kitType Type
hi def link kitStorageClass StorageClass
hi def link kitStructure Structure
hi def link kitTypedef Typedef

" hi def link kitSpecial Special
" hi def link kitSpecialChar SpecialChar
" hi def link kitTag Tag
" hi def link kitDelimiter Delimiter
" hi def link kitSpecialComment SpecialComment
" hi def link kitDebug Debug

" hi def link kitUnderlined Underlined
" hi def link kitIgnore Ignore
" hi def link kitError Error
" hi def link kitTodo Todo
