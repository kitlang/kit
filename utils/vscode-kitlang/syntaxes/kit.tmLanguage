<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>fileTypes</key>
    <array>
      <string>kit</string>
    </array>
    <key>name</key>
    <string>Kit</string>
    <key>scopeName</key>
    <string>source.kit</string>
    <key>uuid</key>
    <string>4fc8848b-af63-4ac5-850f-8e6b00a45004</string>
    <key>patterns</key>
    <array>
      <dict>
        <key>match</key>
        <string>\b(import|include|function|implement|specialize|var|using)\b</string>
        <key>name</key>
        <string>keyword.control.kit</string>
      </dict>
      <dict>
        <key>match</key>
        <string>\b(private|public|static|const|inline)\b</string>
        <key>name</key>
        <string>storage.modifier.kit</string>
      </dict>
      <dict>
        <key>match</key>
        <string>\b(return|throw|for|as|in|if|then|else|do|while|match|default|unsafe|implicit|sizeof)\b</string>
        <key>name</key>
        <string>keyword.other.kit</string>
      </dict>
      <dict>
        <key>match</key>
        <string>\b(struct|enum|union|abstract|rule|rules|typedef|trait)\b</string>
        <key>name</key>
        <string>storage.type.kit</string>
      </dict>
      <dict>
        <key>name</key>
        <string>keyword.control.arrow.kit</string>
        <key>match</key>
        <string>=\&gt;</string>
      </dict>
      <dict>
        <key>name</key>
        <string>support.type.kit</string>
        <key>match</key>
        <string>\b[A-Z][a-zA-Z0-9_]*\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>variable.language.kit</string>
        <key>match</key>
        <string>\b(this|Self)\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>invalid.illegal.macro.kit</string>
        <key>match</key>
        <string>\$[A-Za-z0-9_]+\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>invalid.illegal.macro.kit</string>
        <key>begin</key>
        <string>\$\{</string>
        <key>end</key>
        <string>\}</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.decimal.kit</string>
        <key>match</key>
        <string>\b-?[0-9](\.[0-9+])?+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.octal.kit</string>
        <key>match</key>
        <string>\b-?0o[0-7]+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.hex.kit</string>
        <key>match</key>
        <string>\b-?0x[0-9a-f]+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.bin.kit</string>
        <key>match</key>
        <string>\b-?0b[01]+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>string.quoted.block.kit</string>
        <key>begin</key>
        <string>"""</string>
        <key>end</key>
        <string>"""</string>
      </dict>
      <dict>
        <key>name</key>
        <string>string.quoted.double.kit</string>
        <key>match</key>
        <string>"(\.|[^"])*"</string>
      </dict>
      <dict>
        <key>name</key>
        <string>string.quoted.single.kit</string>
        <key>match</key>
        <string>'(\.|[^\'])*'</string>
      </dict>
      <dict>
        <key>name</key>
        <string>comment.line.double-slash.kit</string>
        <key>match</key>
        <string>(//).*$
?</string>
      </dict>
      <dict>
        <key>name</key>
        <string>comment.block.kit</string>
        <key>begin</key>
        <string>/\*</string>
        <key>end</key>
        <string>\*/</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.language.kit</string>
        <key>match</key>
        <string>\b(true|false|null)\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>markup.heading.metadata.kit</string>
        <key>begin</key>
        <string>\#\[</string>
        <key>end</key>
        <string>\]</string>
      </dict>
      <dict>
        <key>name</key>
        <string>variable.parameter.kit</string>
        <key>match</key>
        <string>\b[a-z_][A-Za-z0-9_]*\b</string>
      </dict>
    </array>
  </dict>
</plist>
