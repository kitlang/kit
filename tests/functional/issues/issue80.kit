import kit.map;
import kit.array;

trait Component{
    function type(): CString;
}

struct PositionComponent{
    public var x: Int;
    public var y: Int;
    public var z: Int;
}

implement Component for PositionComponent{
    function type(): CString{
        return "Position";
    }
}

function main() {

    var myMap: Map[CString, Box[Component]] = Map.new(10);

    var myPosStruct = struct PositionComponent{
        x: 12,
        y: 12,
        z: 23
    };

    var myPosStructBox: Box[Component] = myPosStruct;
    myMap.put(myPosStructBox.type(), myPosStructBox);

    var keys = myMap.keys();
    var constructed = CString.alloc("Position".length);
    var i = 0;
    for char in "Position" {
        constructed[i++] = char;
    }
    printf("key: %s\n", constructed);

    printf("box.type(): %s\n", if myMap.exists(myPosStructBox.type()) then "true" else "false");
    printf("keys[0]: %s\n", if myMap.exists(keys[0]) then "true" else "false");
    printf("literal: %s\n", if myMap.exists("Position") then "true" else "false");
    printf("buffer: %s\n", if myMap.exists(constructed) then "true" else "false");
}
