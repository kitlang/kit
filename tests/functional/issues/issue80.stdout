key: Position
box.type(): true
keys[0]: true
literal: true
buffer: true
